shiny/ui/_theme.py

```diff
@@ -0,0 +1,280 @@
+          addition to Shiny's built-in Bootstrap theme.
+        * `"all"`: Shiny's built-in Bootstrap theme is completely replaced by the theme.
+          This option is for expert usage and should be used with caution. Shiny is
+          designed to work with the currently bundled version of Bootstrap. Use the
+          `bs_version` parameter to check compatibility of the provided theme with the
+          bundled Bootstrap version at runtime.
+
+    Attributes
+    ----------
+    theme
+        The theme to apply as a tagified :class:`~htmltools.Tag`,
+        :class:`~htmltools.TagList`, or :class:`~htmltools.HTMLDependency`.
+    name
+        An optional name for the theme.
+    version
+        An optional version of the theme.
+    bs_version
+        The Bootstrap version with which the theme is compatible.
+    replace
+        Specifies how the theme should replace or augment Shiny's bundled Bootstrap
+        theme:
+
+        * `"css"` is the default: The theme replaces only the `bootstrap.min.css` file
+          of Shiny's built-in Bootstrap theme.
+        * `"none"`: The theme is added as additional CSS (and/or JavaScript) files in
+          addition to Shiny's built-in Bootstrap theme.
+        * `"all"`: Shiny's built-in Bootstrap theme is completely replaced by the theme.
+          This option is for expert usage and should be used with caution. Shiny is
+          designed to work with the currently bundled version of Bootstrap. Use the
+          `bs_version` parameter to check compatibility of the provided theme with the
+          bundled Bootstrap version at runtime.
+
+    Raises
+    ------
+    ValueError
+        If an invalid replacement strategy is specified for `replace`.
+    TypeError
+        If the theme, when tagified, does not return a `Tag` or `TagList`.
+    """
+
+    def __init__(
+        self,
+        theme: str | Path | Tag | Tagifiable | HTMLDependency,
+        *,
+        name: Optional[str] = None,
+        version: Optional[str | Version] = None,
+        bs_version: Optional[str | Version] = None,
+        replace: Literal["css", "all", "none"] = "css",
+    ) -> None:
+        if replace not in ("css", "all", "none"):
+            raise ValueError(
+                f"Invalid value for `replace`: '{replace}'. "
+                + "Must be one of 'css', 'all', 'none'."
+            )
+
+        if isinstance(theme, (str, Path)):
+            theme_tag = head(include_css(theme))
+        elif isinstance(theme, HTMLDependency):
+            theme_tag = theme
+        else:
+            theme_tag = theme.tagify()
+
+            if not isinstance(theme_tag, (Tag, TagList)):
+                raise TypeError(
+                    f"Invalid tagified type for `theme`: {type(theme_tag)}. "
+                    + "When tagified, `theme` must return a `Tag` or `TagList`."
+                )
+
+        self.theme: Tag | TagList | HTMLDependency = theme_tag
```

Comment:
Probably worth commenting that `Tag | TagList` are here to support the `ui.include_css()` case